src/timesignature.ts

```diff
@@ -0,0 +1,137 @@
+// representation
+
+import { Vex } from './vex';
+import { Glyph } from './glyph';
+import { StaveModifier } from './stavemodifier';
+import { TimeSignatureGlyph } from './timesigglyph';
+
+export interface TimeSignatureInfo {
+  glyph: Glyph;
+  line?: number;
+  num: boolean;
+}
+
+const assertIsValidFraction = (timeSpec: string) => {
+  const numbers = timeSpec.split('/').filter((number) => number !== '');
+
+  if (numbers.length !== 2) {
+    throw new Vex.RERR(
+      'BadTimeSignature',
+      `Invalid time spec: ${timeSpec}. Must be in the form "<numerator>/<denominator>"`
+    );
+  }
+
+  numbers.forEach((number) => {
+    if (isNaN(Number(number))) {
+      throw new Vex.RERR('BadTimeSignature', `Invalid time spec: ${timeSpec}. Must contain two valid numbers.`);
+    }
+  });
+};
+
+export class TimeSignature extends StaveModifier {
+  point?: number;
+  bottomLine?: number;
+  protected timeSig?: TimeSignatureInfo;
+
+  protected validate_args: boolean;
+
+  topLine?: number;
+
+  static get CATEGORY(): string {
+    return 'timesignatures';
+  }
+
+  static get glyphs(): Record<string, { code: string; point: number; line: number }> {
+    return {
+      C: {
+        code: 'timeSigCommon',
+        point: 40,
+        line: 2,
+      },
+      'C|': {
+        code: 'timeSigCutCommon',
+        point: 40,
+        line: 2,
+      },
+    };
+  }
+
+  constructor(timeSpec?: string, customPadding = 15, validate_args = true) {
+    super();
+    this.setAttribute('type', 'TimeSignature');
+    this.validate_args = validate_args;
+
+    if (!timeSpec) return;
+    const padding = customPadding;
+
+    this.point = this.musicFont.lookupMetric('digits.point');
+    const fontLineShift = this.musicFont.lookupMetric('digits.shiftLine', 0);
+    this.topLine = 2 + fontLineShift;
+    this.bottomLine = 4 + fontLineShift;
+    this.setPosition(StaveModifier.Position.BEGIN);
+    this.timeSig = this.parseTimeSpec(timeSpec);
+    this.setWidth(this.getTimeSig().glyph.getMetrics().width!);
+    this.setPadding(padding);
+  }
+
+  getCategory(): string {
+    return TimeSignature.CATEGORY;
+  }
+
+  parseTimeSpec(timeSpec: string): TimeSignatureInfo {
+    if (timeSpec === 'C' || timeSpec === 'C|') {
+      const { line, code, point } = TimeSignature.glyphs[timeSpec];
+      return {
+        line,
+        num: false,
+        glyph: new Glyph(code, point),
+      };
+    }
+
+    if (this.validate_args) {
+      assertIsValidFraction(timeSpec);
+    }
+
+    const [topDigits, botDigits] = timeSpec.split('/').map((number) => number.split(''));
+
+    return {
+      num: true,
+      glyph: this.makeTimeSignatureGlyph(topDigits, botDigits),
+    };
+  }
+
+  makeTimeSignatureGlyph(topDigits: string[], botDigits: string[]): Glyph {
+    const glyph = new TimeSignatureGlyph(this, topDigits, botDigits, 'timeSig0', this.point!);
+    return glyph;
+  }
+
+  getTimeSig(): TimeSignatureInfo {
+    return this.timeSig!;
```

Comment:
Better if this returns `TimeSignatureInfo?` and remove `!` from body. The callers can do the explicit check.

Reply:
Actually, looks like `this.timeSig` is set in the constructor, so no need for the protected instance variable to be optional, no?

Reply:
Changed to return `TimeSignature | undefined`, I agree that this is better.
`this.timeSig` is not always set in the constructor due to the `if (!timeSpec) return;`